File: aspnet-in-practice/08-09/CS/CoolMVBlog/Controllers/HomeController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using System.Web.Mvc;
using CoolMvcBlog.Models;
using CoolMvcBlog.Utils;
using System.Web.UI;
using Microsoft.Web.Mvc.AspNet4;
using System.Web.Caching;
using CoolMvcBlog.Utils.Filters;

namespace CoolMvcBlog.Controllers
{
    public class HomeController : Controller
    {
        [LoadTagCloud]
        public ActionResult Index()
        {
            using (var ctx = new BlogModelContainer())
            {
                var lastPosts = ctx.PostSet.OrderByDescending(p => p.DatePublished).Take(3).ToList();
                
                return View(new HomepageModel() { Posts = lastPosts });
            }
        }

        [HttpGet]
        [DependencyOutputCache(Duration = 30, 
            Location=OutputCacheLocation.Server, VaryByParam="None", ParameterName="id")]
        public ActionResult Post(int id, string uniqueName)
        {
            using (var ctx = new BlogModelContainer())
            {
                var post = ctx.PostSet.Include("Comments").Where(p => p.Id == id).Single();
                
                if (post.GetUniqueName() != uniqueName)
                {
                    return new PermanentRedirectResult(
                        this.RedirectToAction("Post", 
                        new { id = id, uniqueName = post.GetUniqueName() }));
                }

                return View(post);
            }
        }

        [HttpPost]
        [RemoveCached(ParameterName = "id")]
        public ActionResult Post(int id, Comment newComment)
        {
            using (var ctx = new BlogModelContainer())
            {
                var post = ctx.PostSet.Include("Comments").Where(p => p.Id == id).Single();

                if (!this.ModelState.IsValid)
                {
                    return this.View(post);
                }

                newComment.Date = DateTime.Now;
                post.Comments.Add(newComment);
                ctx.SaveChanges();

                return RedirectToAction(
                    "Post", 
                    new { id = id, uniqueName = post.GetUniqueName() });
            }
        }
    }
}

File: aspnet-in-practice/08-09/VB/CoolMVCBlog/Controllers/HomeController.vb
Public Class HomeController
    Inherits Controller

    <LoadTagCloud()>
    Public Function Index() As ActionResult
        Using ctx As New BlogModelContainer
            Dim lastPosts = ctx.PostSet.
                OrderBy(Function(p) p.DatePublished).
                Take(3).
                ToList()

            Return View(
                New HomepageModel With {.Posts = lastPosts})
        End Using
    End Function

    <HttpGet()>
    <DependencyOutputCache(Duration:=30,
            Location:=OutputCacheLocation.Server, VaryByParam:="None", ParameterName:="id")>
    Public Function Post(ByVal id As Integer, ByVal uniqueName As String) As ActionResult
        Using ctx = New BlogModelContainer
            Dim thePost = ctx.PostSet.
                Include("Comments").
                Where(Function(p) p.Id = id).
                Single()

            If thePost.GetUniqueName() <> uniqueName Then
                Return New PermanentRedirectResult(Me.RedirectToAction(
                    "Post", New With {.id = id, .uniqueName = thePost.GetUniqueName()}))
            End If

            Return View(thePost)
        End Using
    End Function

    <HttpPost()>
    <RemoveCached(ParameterName:="id")>
    Public Function Post(ByVal id As Integer, ByVal newComment As Comment) As ActionResult
        Using ctx = New BlogModelContainer()
            Dim thePost = ctx.PostSet.
                Include("Comments").
                Where(Function(p) p.Id = id).
                Single()

            If Not Me.ModelState.IsValid Then
                Return Me.View(thePost)
            End If

            newComment.Date = DateTime.Now
            thePost.Comments.Add(newComment)
            ctx.SaveChanges()

            Return Me.RedirectToAction(
                "Post", New With {.id = id, .uniqueName = thePost.GetUniqueName()})
        End Using
    End Function

End Class